old-tutorials-in-development/tutorials-in-development/R/dc-spatio-temporal-intro/09-tabular-data-bad-missing-values.Rmd

```diff
@@ -0,0 +1,361 @@
+
+
+### Check for NA values
+We can quickly check for `NoData` values in our data using the`is.na()`
+function. By asking for the `sum()` of `is.na()` we can see how many NA / missing
+values we have in a dataframe.
+
+```{r NA-values}
+
+# set working directory to ensure R can find the file we wish to import and where 
+# we want to save our files. Be sure to move the unzipped download into your working directory! 
+wd <- "~/Documents/data/" #This will depend on your local environment 
+setwd(wd)
+
+# read csv
+vegData <- read.csv(paste0(wd,"NEON-DS-Field-Site-Spatial-Data/SJER/VegetationData/D17_2013_vegStr.csv"))
+
+# check for NA values in the dbh column
+sum(is.na(vegData$dbh))
+
+```
+
+The result above tells us there are 74 `NoData` values in the `dbh` column.
+
+### Deal with NoData Values
+When we encounter `NoData` values (blank, NAN, -9999, etc.) in our data we need 
+to decide how to deal with them. By default `R` treats `NoData` values designated 
+with a `NA` as a missing value rather than a zero. This is good, as a value of
+zero (no rain today) is not the same as missing data (e.g. we didn't measure the
+amount of rainfall today).
+
+How we deal with `NoData` values will depend on:
+
+* the data type we are working with
+* the analysis we are conducting
+* the significance of the gap or missing value
+
+First, we can see where the `NA` values are in the data set.
+
+``` {r NA-rows, eval=FALSE}
+
+# find the rows with NA values
+vegData[is.na(vegData$dbh),]
+
+```
+
+We can see what rows have `NA` values for `dbh` using the function above. From this 
+information, we may decide to take out rows that contain the `NA` value using 
+the `na.omit` function. Note that this removes **any** row that contains an `NA`. 
+
+``` {r remove-NA-rows}
+
+# create a new data frame without rows containing NAs
+vegDatanoNA <- na.omit(vegData)
+
+```
+
+However, the rows with `NA` values that we just removed do contain good data, 
+so we might not always want to remove the entire row. In this case, we can use 
+`na.rm = TRUE` to remove `NA` values from mathematical operations.
+
+<i class="fa fa-star"></i> **Data Tip:** The functions `mean()`, `median()`, 
+`var()`, `sd()`, `min()` and `max()` all take the `na.rm` argument. The default 
+of `na.rm` is `FALSE` which means if any `NA` is encountered in the data, the 
+output is going to be `NA`. 
+
+``` {r NA-remove}
+
+# mean dbh without removing NAs
+mean(vegData$dbh)
+
+# try again but remove the NAs this time
+mean(vegData$dbh, na.rm = TRUE)
+
+# compare with mean from the new data frame with NAs removed
+mean(vegDatanoNA$dbh)
+
+```
+
+<div id="ds-challenge" markdown="1">
+### Challenge: Mathematical differences
+
+Notice the slight difference in the mean calculations above. 
+
+* What do you think accounts for this difference? 
+* Why is it important to think about potential differences like this?
+
+</div>
+
+``` {r challenge-code-mathematical-differences, echo=FALSE}
```

Comment:
I like the way that you added this challenge code here - fits in with the story and offers a good explanation